Create User Route

RamMaths opened the pull request:
# Create UserProfile Route #16 

- I've added the `UserProfileRoute.tsx` file inside `frontend/src/components` to get the id from url parameters and render the UserProfile component.
- I updated `main.tsx` to add the new route
- I updated `UserProfile.tsx` to load the `shortId` and it is validated as follows

```jsx
// UserProfile.tsx
export function UserProfile({ initialId = '' }: { initialId?: string }) {
  const [shortId, setShortId] = useState<string>(initialId);

  const isValidId = shortId && shortId.trim().length > 0;

  const { loading, error, data } = useQuery(GET_USER, {
    variables: { shortId },
    skip: !isValidId,
  });
  ...
```
- If we get an invalid `shortId` we show the following alert

```jsx
// UserProfile.tsx
{shortId && !isValidId && <Alert type="error">Please enter a valid user ID.</Alert>}
```

## Summary by Sourcery

Add user profile route and enhance frontend routing to support dynamic user profile pages

New Features:
- Create a new route for rendering user profiles with dynamic ID parameter
- Add support for loading user profiles from URL parameters

Enhancements:
- Improve UserProfile component to handle initial ID and validate user ID input
- Update deployment scripts to support frontend deployment

CI:
- Update GitHub Actions workflow to deploy frontend environments
- Modify deployment scripts to support frontend-specific deployment steps

sebastianmarines commented on frontend/src/main.tsx:
```diff
@@ -2,11 +2,13 @@ import { createRoot } from 'react-dom/client';
 import './index.css';
 import App from './App.tsx';
 import { BrowserRouter, Route, Routes } from 'react-router';
+import UserProfileRoute from './components/UserProfileRoute.tsx';
```
```suggestion
import UserProfileRoute from './components/UserProfileRoute.tsx';
import { ApolloProvider } from '@apollo/client';
import { client } from './apollo-client';
```

sebastianmarines commented on frontend/src/main.tsx:
```diff
@@ -2,11 +2,13 @@ import { createRoot } from 'react-dom/client';
+import UserProfileRoute from './components/UserProfileRoute.tsx';
 
 createRoot(document.getElementById('root')!).render(
   <BrowserRouter>
```
```suggestion
<ApolloProvider client={client}>
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<App />} />
        <Route path="/uid/:id" element={<UserProfileRoute />} />
      </Routes>
    </BrowserRouter>
  </ApolloProvider>
```

sebastianmarines commented:
Preview URL: https://dop18dsqjviwb.cloudfront.net/

sebastianmarines reviewed (changes requested):
```
Uncaught Invariant Violation: An error occurred! For more details, see the full error text at https://go.apollo.dev/c/err#%7B%22version%22%3A%223.13.3%22%2C%22message%22%3A58%2C%22args%22%3A%5B%5D%7D
    InvariantError2 invariant.ts:13
```

> Could not find "client" in the context or passed in as an option. Wrap the root component in an `<ApolloProvider>`, or pass an ApolloClient instance in via options.

RamMaths commented:
I corrected the change suggested. Also I added a new component `Layout.tsx` and `main.tsx` has been modified as follows to share the layout of the page with children components:

```tsx
// main.tsx
...
<Routes>
  <Route path="/" element={<Layout />}>
    <Route index element={<App />} />
    <Route path="/uid/:id" element={<UserProfileRoute />} />
  </Route>
</Routes>
...
```

preview URL: https://d3u51inbad6957.cloudfront.net/
